external/fv3fit/fv3fit/sklearn/_min_max_novelty_detector.py

```diff
@@ -0,0 +1,161 @@
+import io
+import joblib
+import logging
+import numpy as np
+from sklearn.preprocessing import MinMaxScaler
+import tensorflow as tf
+import xarray as xr
+import yaml
+
+
+@register_training_function(
+    "min_max_novelty_detector", MinMaxNoveltyDetectorHyperparameters
+)
+def train_min_max_novelty_detector(
+    hyperparameters: MinMaxNoveltyDetectorHyperparameters,
+    train_batches: tf.data.Dataset,
+    validation_batches: tf.data.Dataset,
+):
+    train_batches = train_batches.map(
+        tfdataset.apply_to_mapping(tfdataset.float64_to_float32)
+    )
+
+    return MinMaxNoveltyDetector.fit(train_batches, hyperparameters)
+
+
+@_shared.io.register("minmax")
+class MinMaxNoveltyDetector(NoveltyDetector):
+
+    _PICKLE_NAME = "minmax.pkl"
+    _METADATA_NAME = "metadata.bin"
+
+    scaler: MinMaxScaler
+    input_features_: PackingInfo
+    is_trained: bool = False
+
+    def __init__(self, hyperparameters: MinMaxNoveltyDetectorHyperparameters):
+        super().__init__(hyperparameters.input_variables)
+        self.packer_config = hyperparameters.packer_config
+
+    @classmethod
+    def fit(
+        self,
+        batches: tf.data.Dataset,
+        hyperparameters: MinMaxNoveltyDetectorHyperparameters,
+    ) -> "MinMaxNoveltyDetector":
+        logger = logging.getLogger("MinMaxNoveltyDetector")
+        logger.info(f"Fitting min-max novelty detector.")
+
+        model = MinMaxNoveltyDetector(hyperparameters)
+
+        batches = batches.map(apply_to_mapping(ensure_nd(2)))
+        batches_clipped = batches.map(clip_sample(model.packer_config.clip))
+        x_dataset, model.input_features_ = pack_tfdataset(
+            batches_clipped, [str(item) for item in model.input_variables]
+        )
+        x_dataset = x_dataset.unbatch()
+
+        total_samples = sum(1 for _ in iter(x_dataset))
+        X: Batch = next(iter(x_dataset.batch(total_samples)))
+
+        model.scaler = MinMaxScaler()
+        model.scaler.fit(X)
+
+        logger.info(f"Min-max novelty detector done fitting.")
+        model.is_trained = True
+        return model
+
+    def predict(self, data: xr.Dataset) -> xr.Dataset:
+        """
+        For each coordinate c, computes a score with the following expression:
+        score(c) = max(0, (c - c_max) / c_max) + max(0, (c_min - c) / c_min).
+        A total scores over a column is taken by taking the maximum score.
+        If the score is greater than 0, then at least one coordinate is out of range.
+        """
+        assert self.is_trained
+
+        stack_dims = [dim for dim in data.dims if dim not in stacking.Z_DIM_NAMES]
+        stacked_data = data.stack({SAMPLE_DIM_NAME: stack_dims})
+        stacked_data = stacked_data.transpose(SAMPLE_DIM_NAME, ...)
+
+        X, _ = pack(
+            stacked_data[self.input_variables], [SAMPLE_DIM_NAME], self.packer_config
+        )
+        scaled_X = self.scaler.transform(X)
+        scores_larger_than_max = np.maximum(scaled_X.max(axis=1) - 1, 0)
```

Comment:
It might be convenient to have another method apart from predict that returns just the scores normalized to the min/max in dataset form without the z dimension collapse. ex. you could see which level(s) were going out of range. Up to you if you think that is useful enough to be added as a method - the same information could be manually calculated with `MinMaxNoveltyDetector.scaler`. If you decide to add that it could also be called first in `predict` and the rest of `predict` could just transform the dataset into the stacked scores.